### ISSUE TITLE
Add system prompts independently of history of messages

### ISSUE THREAD

**Issue by josead** [pydantic/pydantic-ai#531]
Hello! I wanted to know how to make the @agent.system_prompt decorator being called on every run, when there is history of messages added. I'm basically trying to evaluate the dependencies on each run (the dependencies will change independently of the run and I don't want to use a tool for this because is not needed)
Here is how to reproduce:

```
 # where I declare the agent
 chat_agent = Agent(
        'openai:gpt-4o',
        system_prompt="Some prompt",
        deps_type=ChatAgentDeps,
    )
    
   ...
    # where I run the agent on each message from the user received
    @chat_agent.system_prompt
    def get_current_status_of_tasks(ctx: RunContext[ChatAgentDeps]):
          print("System Prompt:", ctx.deps.aa_agent.pending_tasks)
          print("System Prompt:", ctx.deps.aa_agent.done_tasks)
          return f"""Pending Action Items: {ctx.deps.aa_agent.get_pending_tasks()}
      Done Action Items: {ctx.deps.aa_agent.get_done_tasks()}"""

    result = await chat_agent.run(message, deps=ChatAgentDeps(
        aa_agent=session["agent"]
    ), message_history=session.get("history", []))
    
    session['history'] = result.all_messages()
```

The function get_current_status_of_tasks only gets called once.

Am I doing something wrong here? Is this expected? Thanks in advance!

**Comment by HamzaFarhan** [comment]
Yes as of now, `self._sys_parts` is not called if `message_history` is passed:

https://github.com/pydantic/pydantic-ai/blob/629495cd1083b74680858f0ce44cc8ba9e19e48a/pydantic_ai_slim/pydantic_ai/agent.py#L835C1-L847C24

```python
async def _prepare_messages(
    self, user_prompt: str, message_history: list[_messages.ModelMessage] | None, run_context: RunContext[AgentDeps]
) -> list[_messages.ModelMessage]:
    if message_history:
        # shallow copy messages
        messages = message_history.copy()
        messages.append(_messages.ModelRequest([_messages.UserPromptPart(user_prompt)]))
    else:
        parts = await self._sys_parts(run_context)
        parts.append(_messages.UserPromptPart(user_prompt))
        messages: list[_messages.ModelMessage] = [_messages.ModelRequest(parts)]


    return messages
```

**Comment by josead** [comment]
Hey @HamzaFarhan, thanks for your quick answer. Is this to prevent the parts to repeat? Is there a simpler way to inject a system prompt into messages? I could do workarounds here, to accept system events to be injected as part of the conversation, do you see the Agents supporting this in the future? 

Regards

**Comment by JacobDu** [comment]
I encountered the same issue in the new version. In my usage scenario, I have two Agents, A and B, each using different strategies to generate system prompts. I need to share the conversation context between the two Agents through message_history. At the same time, I will compress the all_messages in the result and exclude the system prompt message through program logic. 

I hope that when the incoming message_history does not contain a system prompt, the agent can add the current agent's system prompt.
(this seems to be the behavior in version 0.0.12).

**Comment by sydney-runkle** [comment]
Yep, seems reasonable to support this! PRs welcome :)

**Comment by josead** [comment]
Hey @JacobDu ! I have achieved this by modifying the conversation array myself. Maybe what we can add here are utility metods to manipulate "messages". Append, prepend, remove and maybe even go functional here with reduce can be usefull. What do you think?

**Comment by sydney-runkle** [comment]
Closed by https://github.com/pydantic/pydantic-ai/pull/560
